src/EditorFeatures/CSharp/DecompiledSource/AssemblyResolver.cs

```diff
@@ -12,60 +14,117 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.DecompiledSource
     internal class AssemblyResolver : IAssemblyResolver
     {
         private readonly Compilation parentCompilation;
-        private static readonly Version zeroVersion = new Version(0, 0, 0, 0);
+        private readonly Dictionary<string, List<IAssemblySymbol>> cache = new Dictionary<string, List<IAssemblySymbol>>();
+        private readonly StringBuilder logger;
 
-        public AssemblyResolver(Compilation parentCompilation)
+        public AssemblyResolver(Compilation parentCompilation, StringBuilder logger)
         {
             this.parentCompilation = parentCompilation;
+            this.logger = logger;
+            BuildReferenceCache();
+            Log("{0} items in cache", cache.Count);
         }
 
         public PEFile Resolve(IAssemblyReference name)
         {
-            foreach (var assembly in parentCompilation.GetReferencedAssemblySymbols())
+            Log("------------------");
+            Log("Resolve: {0}", name.FullName);
+
+            // First, find the correct list of assemblies by name
+            if (!cache.TryGetValue(name.Name, out var assemblies))
+            {
+                Log("Could not find by name: {0}", name.FullName);
+                return null;
+            }
+
+            // If we have only one assembly available, just use it.
+            // This is necessary, because in most cases there is only one assembly,
+            // but still might have a version different from what the decompiler asks for.
+            if (assemblies.Count == 1)
             {
-                // First, find the correct IAssemblySymbol by name and PublicKeyToken.
-                if (assembly.Identity.Name != name.Name
-                    || !assembly.Identity.PublicKeyToken.SequenceEqual(name.PublicKeyToken ?? Array.Empty<byte>()))
+                Log("Found single assembly: {0}", assemblies[0]);
+                if (assemblies[0].Identity.Version != name.Version)
                 {
-                    continue;
+                    Log("WARN: Version mismatch!");
+                    Log("WARN: Expected: {0}, Got: {1}", name.Version, assemblies[0].Identity.Version);
                 }
+                return MakePEFile(assemblies[0]);
+            }
+
+            // There are multiple assemblies
+            Log("Found {0} assemblies for {1}:", assemblies.Count, name.Name);
```

Comment:
📝 When these are moved to the localization files, the placeholders should be surrounded by `'` characters

```suggestion
            Log("Found '{0}' assemblies for '{1}':", assemblies.Count, name.Name);
```